### PR TITLE
Competition agent integration with zoo

### PR DESCRIPTION
**Pull request by AisenGinn**
competition_agent integration with zoo.

run:`scl run --envision competition_agent_demo.py scenarios/sumo/loop` to see competition integration with loop scenarios. To see other scenarios, replace `scenarios/sumo/loop` with other scenarios path.

### CODE REVIEW

File: zoo/policies/__init__.py [hunk 1]
```diff
@@ -100,3 +104,86 @@ def human_keyboard_entrypoint(*arg, **kwargs):
 
 
 register(locator="human-in-the-loop-v0", entry_point=human_keyboard_entrypoint)
+
+
+from smarts.env.multi_scenario_env import resolve_agent_interface
```
Comment (saulfield) [inline zoo/policies/__init__.py:109]: Is there a reason this isn't at the top of the file? If not, it should be moved.

File: zoo/policies/__init__.py [hunk 2]
```diff
@@ -100,3 +104,86 @@ def human_keyboard_entrypoint(*arg, **kwargs):
...
+root_path = str(Path(__file__).absolute().parents[2])
+
+register(
+    "competition_agent-v0",
+    entry_point=competition_entry,
+    policy_path=os.path.join(root_path, "competition/track1/submission_1"),
+)
+
+register(
+    "competition_agent-v1",
```
Comment (saulfield) [inline zoo/policies/__init__.py:186]: I wonder if these should have distinct names, and then have `v0` for both.

Comment (AisenGinn) [reply zoo/policies/__init__.py:186]: The competition_agent-v1 will be removed in the end after testing some competition policies.

File: zoo/policies/competition_agent.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                    "install",
+                    "-t",
+                    self._sub_env_path,
+                    "-r",
+                    req_file,
+                ]
+            )
+            sys.path.append(self._sub_env_path)
+        except:
+            print("Failed to install requirement for Competition Agent")
```
[the remark this answers is not in the document]
Comment (AisenGinn) [reply zoo/policies/competition_agent.py:42]: Sure, that's reasonable, should I use sys.exit to print the message and exit the program?

File: smarts/zoo/agent_spec.py [hunk 4]
```diff
@@ -74,6 +74,8 @@ class AgentSpec:
     """An adaptor that allows shaping of the reward (default lambda obs, reward: reward)"""
     info_adapter: Callable = lambda obs, reward, info: info
     """An adaptor that allows shaping of info (default lambda obs, reward, info: info)"""
+    adapt_env: Optional[Callable]= lambda env: env
+    """And adaptor that allows configurating the env (default lambda env:env)"""
```
Comment (Adaickalavan) [inline smarts/zoo/agent_spec.py:78]: Adding more adapters counters efforts to remove the adapters #1174 .

**Comment by Adaickalavan** [comment]
+ Why is this pull request being merged into `comp-1` branch?
+ We may grandfather `comp-1` branch and add refactored competition code into `develop` branch such that it can run any external users' agents in general.

**Comment by AisenGinn** [comment]
> * Why is this pull request being merged into `comp-1` branch?
> * We may grandfather `comp-1` branch and add refactored competition code into `develop` branch such that it can run any external users' agents in general.

Currently only comp-1 branch has the competition folder and the require policies/requirements. My work is based on that and merging directly into develop may cause some issues I think.

**Comment by Gamenot** [comment]
> > * We may grandfather `comp-1` branch and add refactored competition code into `develop` branch such that it can run any external users' agents in general.
> 
> Currently only comp-1 branch has the competition folder and the require policies/requirements. My work is based on that and merging directly into develop may cause some issues I think. Maybe we should try to merge comp-1 into develop first?

It is possible that comp-1 does not get merged in. This agent, however, lets us run competition agents directly as were created in the competition so that we will not have to modify them.

File: zoo/policies/competition_agent.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                    "install",
+                    "-t",
+                    self._sub_env_path,
+                    "-r",
+                    req_file,
+                ]
+            )
+            sys.path.append(self._sub_env_path)
+        except:
+            sys.exit("Failed to install requirement for Competition Agent")
```
Comment (Gamenot) [inline zoo/policies/competition_agent.py:43]: Is this necessary, wouldn't the original exception be enough?

```suggestion
        except:
            logging.error("Failed to install requirement for Competition Agent")
            raise
```

File: zoo/policies/competition_agent.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        self._policy_dir = policy_path
+        self._comp_env_path = str(os.path.join(root_path, "competition_env"))
+        self._sub_env_path = str(os.path.join(self._comp_env_path, env_name))
+
+        Path.mkdir(Path(self._comp_env_path), exist_ok=True)
+
+        if Path(self._sub_env_path).exists():
+            shutil.rmtree(self._sub_env_path)
+        Path.mkdir(Path(self._sub_env_path))
```
Comment (Gamenot) [inline zoo/policies/competition_agent.py:26]: This should not be necessary since you can use `parents=True` on line 26.
```suggestion
        if Path(self._sub_env_path).exists():
            shutil.rmtree(self._sub_env_path)
        Path.mkdir(Path(self._sub_env_path), parents=True)
```

**Review by Gamenot — CHANGES_REQUESTED** [review]

File: zoo/policies/__init__.py [hunk 7]
```diff
@@ -100,3 +108,126 @@ def human_keyboard_entrypoint(*arg, **kwargs):
...
+        logging.error(
+            f"Failed to install requirement for Competition Agent in folder {Path(policy_path).name}"
+        )
+        raise
+
+    # insert submission path
+    while policy_path in sys.path:
+        sys.path.remove(policy_path)
+
+    sys.path.insert(0, policy_path)
```
Comment (Gamenot) [inline zoo/policies/__init__.py:166]: Why is this necessary? Is there some sort of leak?
```suggestion
    if policy_path not in sys.path:
        sys.path.insert(0, policy_path)
```

File: zoo/policies/__init__.py [hunk 8]
```diff
@@ -100,3 +108,126 @@ def human_keyboard_entrypoint(*arg, **kwargs):
...
+        adapt_env=env_wrapper,
+        agent_builder=CompetitionAgent,
+    )
+
+    # delete competition policy module and remove related path
+    while policy_path in sys.path:
+        sys.path.remove(policy_path)  # prevent duplicate path remain in sys.path
+
+    while sub_env_path in sys.path:
+        sys.path.remove(sub_env_path)  # prevent duplicate path remain in sys.path
```
Comment (Gamenot) [inline zoo/policies/__init__.py:213]: How does this end up in the path?

File: zoo/policies/competition_agent.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        self._policy_dir = policy_path
+        self._comp_env_path = str(os.path.join(root_path, "competition_env"))
+        self._sub_env_path = str(os.path.join(self._comp_env_path, env_name))
+
+        self._policy = policy
+
+    def act(self, obs):
+        return self._policy.act(obs)
+
+    def close(self, remove_all_env=False):
```
Comment (Gamenot) [inline zoo/policies/competition_agent.py:26]: I do not see where this is called.

File: zoo/policies/__init__.py [hunk 10]
```diff
@@ -100,3 +108,126 @@ def human_keyboard_entrypoint(*arg, **kwargs):
...
+            "policy_path": policy_path,
+            "policy": policy,
+        },
+        adapt_env=env_wrapper,
+        agent_builder=CompetitionAgent,
+    )
+
+    # delete competition policy module and remove related path
+    while policy_path in sys.path:
+        sys.path.remove(policy_path)  # prevent duplicate path remain in sys.path
```
Comment (Gamenot) [inline zoo/policies/__init__.py:210]: Why is it necessary to remove duplicates like this?

File: zoo/policies/competition_agent.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+            sys.path.remove(self._policy_dir)
+        for key, module in list(sys.modules.items()):
+            if "__file__" in dir(module):
+                module_path = module.__file__
+                if module_path and (
+                    self._policy_dir in module_path or self._sub_env_path in module_path
+                ):
+                    sys.modules.pop(key)
+        if remove_all_env:
+            shutil.rmtree(self._comp_env_path, ignore_errors=True)
```
Comment (Gamenot) [inline zoo/policies/competition_agent.py:40]: This cleanup seems to be related to the requirements for all agents of this type rather than the agent instance-specific resources. Maybe this can be registered as an `atexit` callback in the entrypoint?

File: zoo/policies/__init__.py [hunk 12]
```diff
@@ -100,3 +108,126 @@ def human_keyboard_entrypoint(*arg, **kwargs):
...
+    # remove all modules related to policy_path
+    for key, module in list(sys.modules.items()):
+        if "__file__" in dir(module):
+            module_path = module.__file__
+            if module_path and (
+                policy_path in module_path or sub_env_path in module_path
+            ):
+                sys.modules.pop(key)
+
+    del policy_module
```
Comment (Gamenot) [inline zoo/policies/__init__.py:224]: Why are the modules removed at this point?

File: zoo/policies/__init__.py (hunk 10 above)
Comment (AisenGinn) [reply zoo/policies/__init__.py:210]: There are some users that insert the submission path inside the policy.py they wrote, some do not. If they insert there will be two submission paths inside sys.path (which is a list that allow duplication). This is to remove all possible policy paths inside sys.path.

File: test_agent.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
+import gym
+from smarts.zoo import registry
+import sys
+
+test_agent = registry.make_agent("zoo.policies:competition_agent-v0")
+test_agent_spec = registry.make("zoo.policies:competition_agent-v0")
```
[the remark this answers is not in the document]
Comment (AisenGinn) [reply test_agent.py:6]: Is it possible to add an optional argument default to false to let the user decide whether they want to keep the spec?

File: zoo/policies/__init__.py [hunk 14]
```diff
@@ -100,3 +108,126 @@ def human_keyboard_entrypoint(*arg, **kwargs):
...
+
+    return spec
+
+
+register(
+    "competition_agent-v0",
+    entry_point=competition_entry,
+    policy_path=os.path.join(root_path, "competition/track1/submission"),
```
[the remark this answers is not in the document]
Comment (AisenGinn) [reply zoo/policies/__init__.py:232]: Will do.

File: zoo/policies/__init__.py (hunk 7 above)
Comment (AisenGinn) [reply zoo/policies/__init__.py:166]: Just want to ensure that policy_path is been searched first in the sys.path to avoid potential issues. If the policy is already in sys.path and there are some other paths contain `policy.py` in front of it, it may cause some import errors.

File: zoo/policies/competition_agent.py (hunk 9 above)
Comment (AisenGinn) [reply zoo/policies/competition_agent.py:26]: It should be called when all works in done inside ` test_agent.py`. I will modify `test_agent.py` to include this.

File: zoo/policies/__init__.py (hunk 8 above)
Comment (AisenGinn) [reply zoo/policies/__init__.py:213]: The sub_env_path is appended when the dependency is installed.

File: zoo/policies/__init__.py (hunk 12 above)
Comment (AisenGinn) [reply zoo/policies/__init__.py:224]: If the user doesn't called competition_agent.close(), this will ensure they don't have import issues when they import the next competition_agent.

File: zoo/policies/competition_agent.py (hunk 11 above)
Comment (AisenGinn) [reply zoo/policies/competition_agent.py:40]: That sound reasonable, will do.

File: test_agent.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Gamenot) [reply test_agent.py:6]: No, I think we just cache the spec if it is not already cached.

File: zoo/policies/__init__.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Gamenot) [reply zoo/policies/__init__.py:232]: I think we will want to either do it by referencing the winning agents like `"competition_aid_v0"`, `"competition_vcr_v0"`, and `"competition_tju-fanta_v0"`. or by removing a default to force `policy_path` to be specified when initializing an agent.

File: zoo/policies/__init__.py (hunk 7 above)
Comment (Gamenot) [reply zoo/policies/__init__.py:166]: Oh, interesting, the imports resolve in the order they occur in `sys.path`? I think I understand now.

File: zoo/policies/__init__.py (hunk 10 above)
Comment (Gamenot) [reply zoo/policies/__init__.py:210]: OK, that makes more sense.

File: zoo/policies/__init__.py (hunk 12 above)
Comment (Gamenot) [reply zoo/policies/__init__.py:224]: Then one final question: Is the policy module not needed anymore such that the object can be deleted?

File: test_agent.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (AisenGinn) [reply test_agent.py:6]: ok

File: zoo/policies/__init__.py (hunk 7 above)
Comment (AisenGinn) [reply zoo/policies/__init__.py:166]: Will do.

File: zoo/policies/__init__.py (hunk 10 above)
Comment (AisenGinn) [reply zoo/policies/__init__.py:210]: Will do.

File: zoo/policies/__init__.py (hunk 12 above)
Comment (AisenGinn) [reply zoo/policies/__init__.py:224]: I think after the policy & submitted wrapper is once been stored by a variable then the module is not needed anymore. Although keeping the policy_module may not affect the import of next agent.
